server/src/organize-imports.ts

```diff
@@ -5,15 +5,20 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import * as lsp from 'vscode-languageserver';
-import * as tsp from 'typescript/lib/protocol';
-import { Commands } from './commands';
+import * as lsp from "vscode-languageserver";
+import * as tsp from "typescript/lib/protocol";
+import { Commands } from "./commands";
+import { CodeActionKind } from "vscode-languageserver";
 
-export function provideOrganizeImports(file: string, context: lsp.CodeActionContext, result: (lsp.Command | lsp.CodeAction)[]): void {
-    if (!context.only || context.only.indexOf(lsp.CodeActionKind.SourceOrganizeImports) === -1) {
-        return;
+export function provideOrganizeImports(
+    response: tsp.OrganizeImportsResponse | undefined
+): Array<lsp.CodeAction> {
+    if (!response) {
+        return [];
     }
-    result.push(lsp.CodeAction.create("Organize Imports",
-        lsp.Command.create('', Commands.ORGANIZE_IMPORTS, file),
-        lsp.CodeActionKind.SourceOrganizeImports));
+    return response.body.map(edit => lsp.CodeAction.create(
+        "Organize imports",
+        lsp.Command.create("", Commands.ORGANIZE_IMPORTS, edit.fileName),
```

Comment:
Is it possible to provide an edit right away instead of the command?

Command requires a round-trip to the server to execute and with the `only` being used to request the code action, there is a very high chance that the user will trigger that code actions.

Reply:
Possible, but I'd prefer to not do this.

We'd need to convert the [typescript server location](https://github.com/microsoft/TypeScript/blob/cec2fda9a53620dc545a2c4d7b0156446ab145b4/lib/protocol.d.ts#L1510-L1531) (`{start: number, end: number}`) to the [LPS location](https://github.com/microsoft/vscode-languageserver-node/blob/d58c00bbf8837b9fd0144924db5e7b1c543d839e/types/src/main.ts#L112-L122), which we need the [full document text to do](https://github.com/theia-ide/typescript-language-server/blob/f43aff23b9eaaae83d11a422fb509667951b2104/server/src/document.ts#L34-L36). At this context, we don't have that and it's not guaranteed that the file has been opened, technically (I suppose someone could request organize imports on an unopened file, since the document URI is the argument).

It'd be possible to try to get the document at the caller of this and pass it optionally into this `provideOrganizeImports` function, but it feels like that'll be more likely to be flaky to me. (a) the main `response` arg is fetched asynchronously and (b) the edits are based on tsserver state and the optional document text I'd pass in would be typescript-language-server state.

Reply:
OK. I think it would be a good improvement to have when the document is open at least but it can be considered later.